File: khalti-android/src/main/java/com/khalti/checkOut/CheckOutPresenter.java
package com.khalti.checkOut;

import androidx.annotation.NonNull;

import com.khalti.checkOut.helper.CheckoutEventListener;
import com.khalti.checkOut.helper.Config;
import com.khalti.checkOut.helper.MerchantPreferencePojo;
import com.khalti.checkOut.helper.PaymentPreference;
import com.khalti.utils.EmptyUtil;
import com.khalti.utils.GuavaUtil;
import com.khalti.utils.Store;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;

import rx.Subscriber;
import rx.subscriptions.CompositeSubscription;

class CheckOutPresenter implements CheckOutContract.Presenter {
    @NonNull
    private final CheckOutContract.View view;
    private CheckOutModel model;
    private CompositeSubscription compositeSubscription;

    CheckOutPresenter(@NonNull CheckOutContract.View view) {
        this.view = GuavaUtil.checkNotNull(view);
        view.setPresenter(this);
        model = new CheckOutModel();
        compositeSubscription = new CompositeSubscription();
    }

    @Override
    public void onCreate() {
        view.setStatusBarColor();
       /* compositeSubscription.add(RxBus.getInstance().register(Event.class, event -> {
            if (event.getTag().equals("close_check_out")) {
                view.closeCheckOut();
            }
        }));*/

        CheckoutEventListener checkoutEventListener = view::closeCheckOut;
        Store.setCheckoutEventListener(checkoutEventListener);

        /*compositeSubscription.add(view.setTryAgainClickListener().subscribe(aVoid -> fetchPreference(Store.getConfig().getPublicKey())));
        view.toggleIndented(false);
        if (view.hasNetwork()) {
            fetchPreference(Store.getConfig().getPublicKey());
        } else {
            view.showIndentedNetworkError();
        }*/

        view.toggleIndented(false);
        Config c = Store.getConfig();
        List<PaymentPreference> types = c.getPaymentPreferences();

        List<PaymentPreference> uniqueList = new ArrayList<>();
        if (EmptyUtil.isNull(types) || EmptyUtil.isEmpty(types)) {
            uniqueList.add(PaymentPreference.EBANKING);
            uniqueList.add(PaymentPreference.WALLET);
        } else {
            uniqueList.addAll(new LinkedHashSet<>(types));
        }

        view.setupViewPager(uniqueList);
        view.setUpTabLayout(uniqueList);
        view.setTabListener();
    }

    @Override
    public void onDestroy() {
        view.dismissAllDialogs();
    }

    @Override
    public void onTabSelected(int position, boolean selected) {
        view.toggleTab(position, selected);
    }

    @Override
    public void fetchPreference(String key) {
        view.toggleIndented(true);
        compositeSubscription.add(model.fetchPreference(key)
                .subscribe(new Subscriber<MerchantPreferencePojo>() {
                    @Override
                    public void onCompleted() {

                    }

                    @Override
                    public void onError(Throwable e) {
                        view.showIndentedError(e.getMessage());
                    }

                    @Override
                    public void onNext(MerchantPreferencePojo preference) {
                        /*view.toggleIndented(false);
                        List<String> types = preference.getSdkPosition();

                        List<String> uniqueList = new ArrayList<>(new LinkedHashSet<>(types));
                        if (uniqueList.size() == 0) {
                            uniqueList.add(PaymentPreference.EBANKING.getValue());
                            uniqueList.add(PaymentPreference.WALLET.getValue());
                        }

                        view.setupViewPager(uniqueList);
                        view.setUpTabLayout(uniqueList);
                        view.setTabListener();*/
                    }
                }));

    }
}

File: khalti-android/src/main/java/com/khalti/checkout/CheckOutPresenter.kt
package com.khalti.checkout

import com.khalti.checkout.helper.CheckoutEventListener
import com.khalti.checkout.helper.Config
import com.khalti.checkout.helper.PaymentPreference
import com.khalti.signal.CompositeSignal
import com.khalti.signal.Signal
import com.khalti.utils.EmptyUtil
import com.khalti.utils.GuavaUtil
import com.khalti.utils.HandlerUtil
import com.khalti.utils.Store

class CheckOutPresenter(view: CheckOutContract.View) : CheckOutContract.Presenter {
    private val view: CheckOutContract.View = GuavaUtil.checkNotNull(view)
    private val compositeSignal = CompositeSignal()
    private var currentPage = 0

    private val searchList = listOf(PaymentPreference.EBANKING.value, PaymentPreference.MOBILE_BANKING.value)

    init {
        view.setPresenter(this)
    }

    override fun onCreate() {
        view.setStatusBarColor()
        Store.setAppPackageName(view.getAppPackageName())
        Store.setCheckoutEventListener(object : CheckoutEventListener {
            override fun closeCheckout() {
                view.closeCheckOut()
            }
        })

        view.toggleLoading(true)

        val uniqueList = onGetPreferenceList(Store.getConfig())
        HandlerUtil.delayedTask(500) {
            view.toggleLoading(false)
            view.setupViewPager(uniqueList)
            view.toggleTitle(uniqueList.size > 1)

            compositeSignal.add(view.setUpTabLayout(uniqueList)
                .connect {
                    onTabSelected(uniqueList, it)
                })

            HandlerUtil.delayedTask(1000) {
                for (i in 0 until uniqueList.size - 1) {
                    compositeSignal.add(view.setPageScrollListener(i)
                        .connect {
                            view.toggleToolbarShadow(it > 0)
                        })
                }
            }

            compositeSignal.add(view.getSearchViewMapInitSignal()
                .connect {
                    view.toggleSearch(uniqueList[currentPage].value, searchList.contains(uniqueList[currentPage].value))
                })

            view.toggleTestBanner(Store.getConfig().publicKey.contains("test_"))
        }
    }

    override fun onDestroy() {
        view.dismissAllDialogs()
        compositeSignal.clear()
    }

    override fun onGetPreferenceList(config: Config): List<PaymentPreference> {
        val types = config.paymentPreferences

        val uniqueList = ArrayList<PaymentPreference>()
        if (EmptyUtil.isNull(types) || EmptyUtil.isEmpty(types)) {
            uniqueList.add(PaymentPreference.KHALTI)
            uniqueList.add(PaymentPreference.EBANKING)
            uniqueList.add(PaymentPreference.MOBILE_BANKING)
            uniqueList.add(PaymentPreference.CONNECT_IPS)
            uniqueList.add(PaymentPreference.SCT)
        } else {
            uniqueList.addAll(LinkedHashSet<PaymentPreference>(types))
        }

        return uniqueList
    }

    override fun onTabSelected(preferences: List<PaymentPreference>, it: Map<String, Any>) {
        currentPage = it.getValue("position") as Int
        view.toggleSearch(preferences[currentPage].value, searchList.contains(preferences[currentPage].value))
    }

    override fun onBackPressed() {
        Store.getConfig().onCancelListener?.onCancel()
    }
}